### PR TITLE
Cape Town | 26-ITP-Jan | Pretty Taruvinga | Sprint 3 | Implement and rewrite test

### PR DESCRIPTION
**Pull request by Pretty548**
## Learners, PR Template

Self checklist

- [x] I have titled my PR with Region | Cohort | FirstName LastName | Sprint | Assignment Title
- [x] My changes meet the requirements of the task
- [x] I have tested my changes
- [x] My changes follow the [style guide](https://curriculum.codeyourfuture.io/guides/reviewing/style-guide/)

## Changelist

This PR implements the required functions and rewrites the existing tests for Sprint 3 in the `1-implement-and-rewrite-tests` directory.

All tests were updated to correctly reflect the expected behaviour, and the implementations were written to make the tests pass.

### CODE REVIEW

**Review by cjyuan — COMMENTED** [review]
Function implementation looks good.

This exercise has a second part "**2 Rewrite tests with Jest**".

File: Sprint-3/1-implement-and-rewrite-tests/rewrite-tests-with-jest/1-get-angle-type.test.js [hunk 1]
```diff
@@ -14,7 +11,31 @@ test(`should return "Acute angle" when (0 < angle < 90)`, () => {
...
+  expect(getAngleType(359)).toEqual("Reflex angle");
+});
+
 // Case 6: Invalid angles
+test(`should return "Invalid angle" when angle is negative`, () => {
+  expect(getAngleType(-1)).toEqual("Invalid angle");
+});
+test(`should return "Invalid angle" when angle is 360 or greater`, () => {
+  expect(getAngleType(360)).toEqual("Invalid angle");
+});
```
Comment (cjyuan) [inline Sprint-3/1-implement-and-rewrite-tests/rewrite-tests-with-jest/1-get-angle-type.test.js:44]: You did good in identifying one of the boundary cases. There are two boundary cases, and -1 is not one of them. Can you also test the other boundary case?

File: Sprint-3/1-implement-and-rewrite-tests/rewrite-tests-with-jest/2-is-proper-fraction.test.js [hunk 2]
```diff
@@ -4,7 +4,31 @@ const isProperFraction = require("../implement/2-is-proper-fraction");
...
+  expect(isProperFraction(3, 2)).toEqual(false);
+});
+
+test("should return false when numerator is negative", () => {
+  expect(isProperFraction(-1, 2)).toEqual(false);
+});
+
+test("should return true when both numerator and denominator are negative and numerator is smaller than denominator", () => {
+  expect(isProperFraction(-1, -2)).toEqual(true);
+});
```
Comment (cjyuan) [inline Sprint-3/1-implement-and-rewrite-tests/rewrite-tests-with-jest/2-is-proper-fraction.test.js]: The test description on line 28 is not quite accurate. 

For proper fractions, the condition is, "the absolute value of denominator is larger than the absolute value of the numerator". So we can create a category for proper fractions and describe it as:
  `should return true when abs(numerator) < abs(denominator)`

and then test all combinations of positive/negative numerator and denominator (for proper fraction) in this category.

**Review by cjyuan — COMMENTED** [review]
Please note that in CYF courses, the recommended way to inform the reviewer of your changes is to do both of the following:
  - **Reply to their feedback.**
    - In the responses, clarify how each piece of feedback was addressed to demonstrate that you've carefully reviewed the suggestions.
      - You may find the suggestions in this [PR Guide](https://programming.codeyourfuture.io/guides/reviewing/trainee-pr-guide/#addressing-reviewer-feedback) useful.
    - Your response may trigger a notification (depending on the reviewer's settings), helping ensure they’re aware of the updates you’ve made. 
  - **Replace the "Reviewed" label by a "Needs review" label** (which you have done -- great!)
    - Without this label, the reviewer would not know if your changes is ready to be reviewed.

File: Sprint-3/1-implement-and-rewrite-tests/rewrite-tests-with-jest/1-get-angle-type.test.js (hunk 1 above)
Comment (Pretty548) [reply Sprint-3/1-implement-and-rewrite-tests/rewrite-tests-with-jest/1-get-angle-type.test.js:44]: Thanks for the feedback! I’ve added a test for the other boundary case (0) and updated the tests.

File: Sprint-3/1-implement-and-rewrite-tests/rewrite-tests-with-jest/2-is-proper-fraction.test.js (hunk 2 above)
Comment (Pretty548) [reply Sprint-3/1-implement-and-rewrite-tests/rewrite-tests-with-jest/2-is-proper-fraction.test.js]: Thank you for the feedback. I’ve updated the tests to follow the rule that a proper fraction is when `abs(numerator) < abs(denominator)`. I grouped the proper fraction cases together and included the different positive and negative combinations. Please let me know if there is anything else I should improve.

File: Sprint-3/1-implement-and-rewrite-tests/rewrite-tests-with-jest/1-get-angle-type.test.js [hunk 3]
```diff
@@ -14,7 +11,37 @@ test(`should return "Acute angle" when (0 < angle < 90)`, () => {
...
+test(`should return "Invalid angle" when angle is 0`, () => {
+  expect(getAngleType(0)).toEqual("Invalid angle");
+});
+test(`should return "Invalid angle" when angle is 360 or greater`, () => {
+  expect(getAngleType(360)).toEqual("Invalid angle");
+});
+test(`should return "Invalid angle" when angle is not a number`, () => {
+  expect(getAngleType("not a number")).toEqual("Invalid angle");
+});
```
Comment (cjyuan) [inline Sprint-3/1-implement-and-rewrite-tests/rewrite-tests-with-jest/1-get-angle-type.test.js]: What do you expect from this function call:

`getAngleType("100")` or `getAngleType("      +1e2")` ? 

Note: You don't have to include this test because the spec has already stated you can assume the parameters are always valid numbers.

File: Sprint-3/1-implement-and-rewrite-tests/rewrite-tests-with-jest/2-is-proper-fraction.test.js [hunk 4]
```diff
@@ -4,7 +4,29 @@ const isProperFraction = require("../implement/2-is-proper-fraction");
...
   expect(isProperFraction(1, 0)).toEqual(false);
 });
+
+// Case: numerator is zero
+test("should return true when numerator is zero and denominator is positive", () => {
+  expect(isProperFraction(0, 1)).toEqual(true);
+});
+
+// Proper fractions
+test("should return true when abs(numerator) < abs(denominator)", () => {
```
Comment (cjyuan) [inline Sprint-3/1-implement-and-rewrite-tests/rewrite-tests-with-jest/2-is-proper-fraction.test.js:18]: You could also apply this notation to improper fractions.

File: Sprint-3/1-implement-and-rewrite-tests/rewrite-tests-with-jest/1-get-angle-type.test.js (hunk 3 above)
Comment (Pretty548) [reply Sprint-3/1-implement-and-rewrite-tests/rewrite-tests-with-jest/1-get-angle-type.test.js]: Thanks for the clarification! I did not include tests for non-number inputs since the spec assumes valid numbers.

File: Sprint-3/1-implement-and-rewrite-tests/rewrite-tests-with-jest/2-is-proper-fraction.test.js (hunk 4 above)
Comment (Pretty548) [reply Sprint-3/1-implement-and-rewrite-tests/rewrite-tests-with-jest/2-is-proper-fraction.test.js:18]: Thanks for the suggestion! I updated the improper fraction test description to use the same abs notation and added a negative numerator/denominator case.

**Comment by cjyuan** [comment]
All good now. Well done.

**Comment by illicitonion** [comment]
Closing PR because the January ITP run has finished. Feel free to re-open if you're still working on it.
